File: src/main/java/com/ferreusveritas/exampletrees/trees/TreeIron.java
package com.ferreusveritas.exampletrees.trees;

import java.util.List;
import java.util.Random;

import com.ferreusveritas.dynamictrees.blocks.BlockDynamicSapling;
import com.ferreusveritas.dynamictrees.systems.dropcreators.DropCreatorSeed;
import com.ferreusveritas.dynamictrees.trees.TreeFamily;
import com.ferreusveritas.dynamictrees.trees.Species;
import com.ferreusveritas.exampletrees.ModBlocks;
import com.ferreusveritas.exampletrees.ModConstants;

import net.minecraft.block.Block;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import net.minecraft.world.biome.Biome;
import net.minecraftforge.common.BiomeDictionary;
import net.minecraftforge.common.BiomeDictionary.Type;

public class TreeIron extends TreeFamily {
	
	//Species need not be created as a nested class.  They can be created after the tree has already been constructed.
	public class TreeIronSpecies extends Species {
		
		public TreeIronSpecies(TreeFamily treeFamily) {
			super(treeFamily.getName(), treeFamily, ModBlocks.ironLeavesProperties);

			//Immensely slow-growing, stocky tree that pulls trace amounts of iron from the dirt
			setBasicGrowingParameters(0.5f, 10.0f, getUpProbability(), getLowestBranchHeight(), 0.1f);
			
			//Setup the dynamic sapling.  This could be done outside of the constructor but here is fine.
			setDynamicSapling(new BlockDynamicSapling("ironsapling").getDefaultState());
			
			//This will allow the tree to grow in the Mesa which has very little traditional dirt.
			addAcceptableSoil(Blocks.STAINED_HARDENED_CLAY, Blocks.HARDENED_CLAY, Blocks.SAND);
			
			//Let's pretend that iron trees have a hard time around water because of rust or something
			envFactor(Type.BEACH, 0.1f);
			envFactor(Type.WET, 0.25f);
			envFactor(Type.WATER, 0.25f);
			envFactor(Type.DRY, 1.05f);
			
			addDropCreator(new DropCreatorSeed(0.1f) {
				@Override
				public List<ItemStack> getHarvestDrop(World world, Species species, BlockPos leafPos, Random random, List<ItemStack> dropList, int soilLife, int fortune) {
					return dropList;
				}
				
				@Override
				public List<ItemStack> getLeavesDrop(IBlockAccess access, Species species, BlockPos breakPos, Random random, List<ItemStack> dropList, int fortune) {
					return dropList;
				}
			});
		}
		
		@Override
		public boolean isBiomePerfect(Biome biome) {
			//Let's pretend that Dry Mesa biomes have a lot of iron in the clays that help these trees grow.
			return BiomeDictionary.hasType(biome, Type.MESA);
		}
		
		/*@Override
		public void addJoCodes() {
			//Disable adding of JoCodes
		}*/
		
	}
		
	public TreeIron() {
		super(new ResourceLocation(ModConstants.MODID, "iron"));
		
		//Set up primitive log. This controls what is dropped on harvest.
		setPrimitiveLog(ModBlocks.ironLog.getDefaultState());

		ModBlocks.ironLeavesProperties.setTree(this);
	}

	@Override
	public void createSpecies() {
		setCommonSpecies(new TreeIronSpecies(this));
		getCommonSpecies().generateSeed();
	}

	//Since we created a DynamicSapling in the common species we need to let it out to be registered.
	@Override
	public List<Block> getRegisterableBlocks(List<Block> blockList) {
		blockList.add(getCommonSpecies().getDynamicSapling().getBlock());
		return super.getRegisterableBlocks(blockList);
	}
	
}

File: src/main/java/com/ferreusveritas/exampletrees/trees/TreeIron.kt
package com.ferreusveritas.exampletrees.trees

import com.ferreusveritas.dynamictrees.blocks.BlockDynamicSapling
import com.ferreusveritas.dynamictrees.systems.dropcreators.DropCreatorSeed
import com.ferreusveritas.dynamictrees.trees.Species
import com.ferreusveritas.dynamictrees.trees.TreeFamily
import com.ferreusveritas.exampletrees.ModConstants
import com.ferreusveritas.exampletrees.categories.ModBlocks
import net.minecraft.block.Block
import net.minecraft.init.Blocks
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import net.minecraft.world.biome.Biome
import net.minecraftforge.common.BiomeDictionary
import net.minecraftforge.common.BiomeDictionary.Type
import java.util.*

class TreeIron : TreeFamily(ResourceLocation(ModConstants.MODID, "iron")) {

    //Species need not be created as a nested class.  They can be created after the tree has already been constructed.
    inner class TreeIronSpecies(treeFamily: TreeFamily) : Species(treeFamily.name, treeFamily, ModBlocks.ironLeavesProperties) {

        init {

            //Immensely slow-growing, stocky tree that pulls trace amounts of iron from the dirt
            setBasicGrowingParameters(0.5f, 10.0f, getUpProbability(), getLowestBranchHeight(), 0.1f)

            //Setup the dynamic sapling.  This could be done outside of the constructor but here is fine.
            dynamicSapling = BlockDynamicSapling("ironsapling").defaultState

            //This will allow the tree to grow in the Mesa which has very little traditional dirt.
            addAcceptableSoil(Blocks.STAINED_HARDENED_CLAY, Blocks.HARDENED_CLAY, Blocks.SAND)

            //Let's pretend that iron trees have a hard time around water because of rust or something
            envFactor(Type.BEACH, 0.1f)
            envFactor(Type.WET, 0.25f)
            envFactor(Type.WATER, 0.25f)
            envFactor(Type.DRY, 1.05f)

            addDropCreator(object : DropCreatorSeed(0.1f) {
                override fun getHarvestDrop(world: World?, species: Species, leafPos: BlockPos?, random: Random, dropList: List<ItemStack>, soilLife: Int, fortune: Int): List<ItemStack> {
                    return dropList
                }

                override fun getLeavesDrop(access: IBlockAccess?, species: Species, breakPos: BlockPos?, random: Random, dropList: List<ItemStack>, fortune: Int): List<ItemStack> {
                    return dropList
                }
            })
        }

        override fun isBiomePerfect(biome: Biome?): Boolean {
            //Let's pretend that Dry Mesa biomes have a lot of iron in the clays that help these trees grow.
            return BiomeDictionary.hasType(biome!!, Type.MESA)
        }

        /*@Override
		public void addJoCodes() {
			//Disable adding of JoCodes
		}*/

    }

    init {

        //Set up primitive log. This controls what is dropped on harvest.
        primitiveLog = ModBlocks.ironLog.getDefaultState()

        ModBlocks.ironLeavesProperties.setTree(this)
    }

    override fun createSpecies() {
        setCommonSpecies(TreeIronSpecies(this))
        getCommonSpecies().generateSeed()
    }

    //Since we created a DynamicSapling in the common species we need to let it out to be registered.
    override fun getRegisterableBlocks(blockList: MutableList<Block>): List<Block> {
        blockList.add(getCommonSpecies().dynamicSapling.block)
        return super.getRegisterableBlocks(blockList)
    }

}
